Lib/test/test_dataclasses.py

```diff
@@ -2902,6 +2902,59 @@ class A:
         self.assertEqual(obj.a, 'a')
         self.assertEqual(obj.b, 'b')
 
+    def test_frozen_slots_raises_correct_error(self):
+        # bpo-45897
+        @dataclass(slots=True, frozen=True)
+        class Point:
+            x: int
+            y: int
+
+        p = Point(1, 2)
+        self.assertEqual(p.x, 1)
+        self.assertEqual(p.y, 2)
+        with self.assertRaises(FrozenInstanceError):
+            p.x = 2
+
+        with self.assertRaises(AttributeError) as ex:
+            p.z = 5
+        type_of_caught_exception = type(ex.exception)
+
+        # Verify that the caught exception is `AttributeError` exactly,
+        # and *not* a subclass of `AttributeError` such as `FrozenInstanceError`.
+        #
+        # We need to ensure that this error is being caused by the standard machinery that
+        # prevents new attributes being added that aren't specified in __slots__.
+        #
+        # We *do not* want the error to be caused by anything related to dataclasses.
+        self.assertIs(
+            type_of_caught_exception, 
+            AttributeError, 
+            msg=f"expected 'AttributeError', got {type_of_caught_exception.__name__!r}"
+        )
```

Comment:
`assertRaisesRegex` will help you differentiate between `AttributeError` and `FrozenInstanceError`.

```suggestion
        regex = "object has no attribute"
        with self.assertRaisesRegex(AttributeError, regex):
            p.z = 5
```

Reply:
Also note that if you change this to `assertRaises(FrozenInstanceError)` it actually fails. The following tests will pass when passed `AttributeError` through.

Reply:
But this version doesn't test the actual type of the exception, just the text. If the "wrong" type of exception had the "right" text, this would pass, in error. Correct?

Reply:
Hmm. Not sure about this. Testing on the phrasing of the exception message feels risky. What if the FrozenInstanceError message changes at some point? Or the standard AttributeError you get with slots? I'd prefer to test the identity of the class, as I'm doing at the moment :)

Reply:
Correct. So this then (I still prefer to keep the regex; it's almost always a good idea):

```suggestion
        regex = "object has no attribute"
        with self.assertRaisesRegex(AttributeError, regex) as cm:
            p.z = 5
        self.assertIsInstance(cm.exception, AttributeError)
```

Reply:
`self.assertIsInstance` will pass if `FrozenInstanceError` is raised instead of `AttributeError`, so it has to be `assertIs` 🙂

Reply:
Ah, you're right! Thanks :)